broker-core/src/main/java/io/zeebe/broker/workflow/state/ElementInstanceState.java

```diff
@@ -206,15 +210,18 @@ private int writeStoreRecordKeyIntoBuffer(
     return offset;
   }
 
-  public void storeRecord(long scopeKey, TypedRecord<WorkflowInstanceRecord> record) {
+  public void storeRecord(
+      long scopeKey, TypedRecord<WorkflowInstanceRecord> record, Purpose purpose) {
     final IndexedRecord indexedRecord =
         new IndexedRecord(
             record.getKey(),
             (WorkflowInstanceIntent) record.getMetadata().getIntent(),
             record.getValue());
 
+    final StoredRecord storedRecord = new StoredRecord(indexedRecord, purpose);
+
     final int keyLength = writeStoreRecordKeyIntoBuffer(keyBuffer, 0, scopeKey, record.getKey());
```

Comment:
Let us use as key [scopekey,purpose,recordKey] then we can search for [scopekey,purpose] as prefix